### ISSUE TITLE
ERROR holochain_websocket::websocket: websocket_error_from_network=Utf8 from Conductor when attempting to call a zome function

### ISSUE THREAD

**Issue by dellams** [holochain/holochain#839]
Greetings,

As you may be aware I built HoloNET back in 2019 to connect Holochain to .NET and Unity clients. I have been trying to upgrade it to RSM and have integrated MessagePack and reversed engineered your nodejs client the best I can.

I have been trying to get this working for many months now with lots of ping/pong on the forum:

https://forum.holochain.org/t/about-the-rsm-technical-discussion-category/4086/153

I am currently getting the error:

`ERROR holochain_websocket::websocket: websocket_error_from_network=Utf8`

I am connecting to the condcutor using WSL2 on Windows 10. I am using the default setup from the DNA build tutorial here:
holochain/holochain-dna-build-tutorial

My default.nix file looks like this:

```
let 
  holonixPath = builtins.fetchTarball {
    url = "https://github.com/holochain/holonix/archive/90a19d5771c069dbcc9b27938009486b54b12fb7.tar.gz";
    sha256 = "11wv7mwliqj38jh1gda3gd0ad0vqz1d42hxnhjmqdp037gcd2cjg";
  };
  holonix = import (holonixPath) {
    includeHolochainBinaries = true;
    holochainVersionId = "custom";
    
    holochainVersion = { 
     rev = "d3a61446acaa64b1732bc0ead5880fbc5f8e3f31";
     sha256 = "0k1fsxg60spx65hhxqa99nkiz34w3qw2q4wspzik1vwpkhk4pwqv";
     cargoSha256 = "0fz7ymyk7g3jk4lv1zh6gbn00ad7wsyma5r7csa88myl5xd14y68";
     bins = {
       holochain = "holochain";
       hc = "hc";
     };
    };
  };
in holonix.main
```

Please see the forum link above for more detail on what I am passing to the conductor. 

Any ideas? Are they are logs I can look at, etc?

Any help would be greatly appreciated.

Keep up the great work, thanks! :)

Many thanks,
David.

**Comment by Connoropolous** [comment]
Further info that would help ... 

Can you show additional detail about the code that is trying to connect via websocket? e.g. the client. 
Is it .NET code? what websocket library are you using? 

At what point does the error message you quoted arise? What is happening just before and just after that?

**Comment by Connoropolous** [comment]
Update: your title says that you see that error when you make a zome call. Each time? 

What library did you utilize for msgpack?

**Comment by dellams** [comment]
Did you not see the code I already shared in the forum link above? :)

![image](https://user-images.githubusercontent.com/3522378/124780300-65908780-df3a-11eb-861d-ecde9fa98a11.png)

Yes it is .net/c# code and it uses my own hand written web socket code (it only uses some of the internal .net websocket classes).

Here is the full code (including Web Socket Code):

https://github.com/NextGenSoftwareUK/Our-World-OASIS-API-HoloNET-HoloUnity-And-.NET-HDK/blob/master/NextGenSoftware.Holochain.HoloNET.Client.Core/HoloNETClientBase.cs

The issue is in the CallZomeFunctionAsync method so it connects fine with the Connect method higher up, but when you then try to call a zome function, that is when you get the Conductor error.

It is using this MessagePack library:
https://github.com/neuecc/MessagePack-CSharp

After doing research, it looked like this was the fastest MessagePack library for .net.

I hope this was all you was after? :) 

Many thanks,
David.

**Comment by Connoropolous** [comment]
Hey @dellams I hadn't seen that code snippet before no, but it is most useful as we have it now to see it in the full context. 

I don't know CSharp, but I know Holochain, so I'll see if I can at least help diagnose the issue, and maybe attract some eyeballs to resolve it.

**Comment by Connoropolous** [comment]
Let's start with a diagnosis, a quick search for the error message "websocket_error_from_network" has one specific search result hit, so that's a strong start.
https://github.com/holochain/holochain/blob/b8222ea979f03c001dadcf33c13113ccd686c6e8/crates/holochain_websocket/src/websocket.rs#L505

That function is called "process_from_websocket" and I guess it already has your Zome call websocket message as a `Result` which is an `Err` state, rather than `Ok` since its taking that `match` pathway. 
And `msg` in the context is already defined when the function is called, so we have to walk further back...

There is only one place that calls `process_from_websocket` too... 
`msg` should be a `std::result::Result<tungstenite::Message, tungstenite::Error>`
and obviously we want it to be an `Ok` with a `tungstenite::Message` not an `Err` with a `tungstenite::Error`

tungstenite is an imported websocket library in use by the crate. we should check the version
https://github.com/holochain/holochain/blob/b8222ea979f03c001dadcf33c13113ccd686c6e8/crates/holochain_websocket/Cargo.toml#L24

A tungstenite message is an enum like this:
https://docs.rs/tungstenite/0.13.0/tungstenite/protocol/enum.Message.html

So in order for your message to work correctly, it would need to serialize in a way that tungstenite expects.
We can look at holochain-conductor-api, the js client, to check how it does it

**Comment by Connoropolous** [comment]
It looks like you've already done that, but there must be something off or different about it... let's try to see...

**Comment by Connoropolous** [comment]
@dellams the only thing I'm picking up here is that it's possible that `id` being a string is breaking this.

https://github.com/holochain/holochain/blob/2d9401a5e0f74934195a0bf02ca198679b53089d/crates/holochain_websocket/src/lib.rs#L138-L167

In holochain-conductor-api `id` is and remains an integer when being serialized:
https://github.com/holochain/holochain-conductor-api/blob/5b72d400dceeba14de258cae6a31c8fe3b81b7cd/src/websocket/client.ts#L76

**Comment by Connoropolous** [comment]
In your struct, `id` is a string:
https://github.com/NextGenSoftwareUK/Our-World-OASIS-API-HoloNET-HoloUnity-And-.NET-HDK/blob/50aa3db3588910b817aed194fc575c3e9900981a/NextGenSoftware.Holochain.HoloNET.Client.Core/HoloNETRequest.cs#L10

**Comment by Connoropolous** [comment]
it's definitely worth a shot!

**Comment by dellams** [comment]
Ok thanks, I changed it to a ulong so it matches the u64 in the hc code. But now unfortunately I am getting the connection refused error message again! Arghhhhhhhhhhhhhhhhhh! lol

Last time I resolved this with the wslconfig workaround as detailed here:
https://forum.holochain.org/t/about-the-rsm-technical-discussion-category/4086/151

But this no longer seems to work.

Any ideas?

Cheers
D.

**Comment by Connoropolous** [comment]
Do you mean this one?

Error Occured. Resason: Error occurred connecting to ws://localhost:8888
Error Details: System.Net.WebSockets.WebSocketException (0x80004005): Unable to connect to the remote server
—> System.Net.Http.HttpRequestException: No connection could be made because the target machine actively refused it.

**Comment by dellams** [comment]
Yes, the bane of my life! lol

**Comment by Connoropolous** [comment]
So you're saying you changed nothing else, but the ulong thing, and suddenly you get this instead?

So you still have this in your .wslconfig? 
```
[wsl2]
localhostForwarding=true
```
and that's not working for you now?

**Comment by Connoropolous** [comment]
@dellams this issue, by the looks of it to me, is totally in the wsl2 area... 
https://github.com/microsoft/WSL/issues/5211#issuecomment-706652801
there are a couple possible workarounds, discussed here. I get the impression that oddly enough just a 'stop and restart' could potentially unblock you. Seems to be a lot of frustration around this

**Comment by Connoropolous** [comment]
Good news though, there is the potential for non-wsl style Windows support for Holochain coming through the pipes!

**Comment by dellams** [comment]
> So you're saying you changed nothing else, but the ulong thing, and suddenly you get this instead?
> 
> So you still have this in your .wslconfig?
> 
> ```
> [wsl2]
> localhostForwarding=true
> ```
> 
> and that's not working for you now?

Yes, but it was not the ulong change that caused the issue, just a coincidence. Yeah your right this is an issue with WSL2, nothing else. I have no idea why the wslconfig fix has stopped working?

I did quite a lot of research into this including the link you shared before I found this fix. I tried lots of other things before but nothing worked except this fix so I don't have anywhere else to go now... :(

Does anyone else in the hc community have this working with WSL2? 

I know why lots of people prefer Linux! ;-) lol I do too but unfortunately I am stuck with Windows until they release VS on Linux, which sadly I am not sure they ever will...

I think I will compile HoloNET RSM and test it on my Linux install, cant make many code changes there, because Mono and VSCode are not up to the task... they are just like text editors compared to the full VS.

**Comment by Connoropolous** [comment]
If youre prepared to change your angle of approach, then you might try the fresh new commits on the develop branch that include native windows compile target for holochain

**Comment by dellams** [comment]
WOW, yes please! :) Finally! :) I have waited 2 years for a native windows build for the conductor! :) I originally built HoloNET against the windows binary of the conductor back in 2019 before NixOS came along... It could auto-start/shutdown the conductor and be deeply integrated into any .net build so people did not ever need to worry about installing/configuring/starting/shutting down the conductor... :) 

So is there now a exe for the holochain conductor again? :)

Cheers
D.

**Comment by Connoropolous** [comment]
@ddd-mtl says that windows support is merged in since the following commit

FYI Windows support has been merged in at REV c5dbdf28825927106bc32d186dd54f20d35df468

You'd have to build holochain from source after that commit, for the windows build target. Would you know how or need some pointers?

**Comment by dellams** [comment]
Yes please, if you have clear step by step instructions that would be really appreciated thanks! :)

I feel this is a VERY BIG deal and big step in the right direction for Holochain so I feel this should be made public and announced. We can at least for starters share the setup guide in the forum? :)

Very HAPPY, THANK YOU! :)

Keep up the great work, I now finally feel more of the Holochain Crew again after feeling a little outcast because I was waiting for the Windows build for nearly 2 years... :) 

I can now finally finish HoloNET RSM, .NET HDK and STAR ODK... :) Thank you.

Time to open the foodgates to Holochain by bringing the massive .net, unity and business enterprise markets to Holochain.... :) 

Much love & blessings,
David. ❤️ ❤️ ❤️

**Comment by Connoropolous** [comment]
I'm excited too but let's save that extra enthusiasm for after we see it working hah, cause we've only had 1 other person utilize this so far. What's the exact details of your windows OS?

**Comment by dellams** [comment]
Windows 10, what else did you need to know? 🙂

Out of interest who else is using this? 🙂

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by dellams** [comment]
Oh is this not meant to be public knowledge yet? Thanks

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by Connoropolous** [comment]
Ok ok gotcha. Well, the background here is that this work to build windows support was an initiative started outside of Holochain core team. It's not 'secret' by any means, but it's not well tested either meaning promoting it loudly could backfire. 
I am planning in the next few days ( I can't today :( ) to spin up our Windows OS and give things a shot. 

It's @ddd-mtl as mentioned a few comments ago, who's done the heavy lifting and been also working things out on their system. 

Windows 10 is good, I think that's the main requirement. Perhaps not even a requirement, but again, untested. There's no CI for Windows support yet either which solving that will help it stabilize and stay stable. 

I am assuming, untested, that the pathway is sometihng like this. 

1. https://rustup.rs/# 
2. do one of these <img width="839" alt="Screen Shot 2021-07-28 at 2 08 43 PM" src="https://user-images.githubusercontent.com/1409121/127396278-364343dc-624e-4baf-88b2-4b80fb4f829b.png">
3. `git clone https://github.com/holochain/holochain`
4. `cd holochain`
5. `cargo install --path crates/holochain`
6. `cargo install --path crates/hc`

**Comment by dellams** [comment]
Oh right, yes, I did try that a few months ago, but ran into a whole set of new errors! ;-)  lol

I posted them on that long thread on the forum, where we first connected, but at the time no-one responded so I had to park it and go back to NixOS way...

But I will give it another try and report back here... :)

Really appreciate all your help and really excited that other guy has got unity working with RSM, gives me hope for HoloNET and someone else I can work with, yay! :)

Only taken 2 lonely years but finally support is arriving, hurray! :)

Cheers
D.

**Comment by dellams** [comment]
By the way, love the work you are doing with Acorn and hope to start using it for my project too... any pointers or steps to get up and running with it? Thanks :)

**Comment by dellams** [comment]
Will need to thank  @ddd-mtl for this great work, looks like he is working on SnapMail? I watched your videos with David Atkinson about that too as I recall, you working on that as well? :)

**Comment by Connoropolous** [comment]
"By the way, love the work you are doing with Acorn and hope to start using it for my project too... any pointers or steps to get up and running with it?"

If this windows native developments turn out to be working smoothly, then a windows build of Acorn will be very close on the horizon as well :) so you're helping right now just by trying!

**Comment by Connoropolous** [comment]
I'm not working on SnapMail directly at all, but Damien and I share notes and pointers often.

**Comment by Connoropolous** [comment]
By the way, it was a sub component, lair keystone, that actually needed Windows compatibility added to it, and thats the work that was done to make this happen. Then lair keystone version needed to be upgraded in terms of the one holochain uses. The first time you run Holochain binary, it will start to automatically download and compile the right windows compatible version of lair keystone. So the first run takes a long time because it is running a cargo install command which needs to compile. Second run will be fast as the right version will be installed.

**Comment by dellams** [comment]
Ok great thanks, yes, that was the error I had last time with the keystones as I recall! :)

Ok, will give it a go and report back... :)

I am still wondering how the other guy with the unity code got around the connection refused issue I am having?

Cheers
D,

**Comment by Connoropolous** [comment]
Let me know how it goes! 

The issue you are having is wsl, which applies in the Windows context. He is on mac.

**Comment by dellams** [comment]
Oh right ok, cool. :)

**Comment by dellams** [comment]
Ok, I ran ` cargo install --path crates/holochain` and got this error:

```
error: failed to compile `holochain v0.0.102 (C:\Users\david\holochain\crates\holochain)`, intermediate artifacts can be found at `C:\Users\david\holochain\target`

Caused by:
  failed to parse manifest at `C:\Users\david\.cargo\registry\src\github.com-1ecc6299db9ec823\cranelift-frontend-0.74.0\Cargo.toml`

Caused by:
  feature `resolver` is required

  this Cargo does not support nightly features, but if you
  switch to nightly channel you can add
  `cargo-features = ["resolver"]` to enable this feature
```

Any ideas? Thanks :)

**Comment by dellams** [comment]
Do I need to install something else first?

**Comment by Connoropolous** [comment]
I think this is something easily fixed. @dellams are you at the comp to go back and forth if I try to help right now?

**Comment by Connoropolous** [comment]
@dellams  I see in a chat (I've blurred names) another talking about this
<img width="815" alt="Screen Shot 2021-08-03 at 10 12 17 AM" src="https://user-images.githubusercontent.com/1409121/128057921-69b5d3a8-a8d6-470d-a7b2-c98a25e317ce.png">

when you do `rustc --version` what's your version? are you knowledgeable about how to upgrade? It's not hard

**Comment by Connoropolous** [comment]
`rustup toolchain install 1.54` might do the trick for you

**Comment by dellams** [comment]
> @dellams I see in a chat (I've blurred names) another talking about this
> <img alt="Screen Shot 2021-08-03 at 10 12 17 AM" width="815" src="https://user-images.githubusercontent.com/1409121/128057921-69b5d3a8-a8d6-470d-a7b2-c98a25e317ce.png">
> 
> when you do `rustc --version` what's your version? are you knowledgeable about how to upgrade? It's not hard

rustc 1.50.0 (cb75ad5db 2021-02-10)

**Comment by dellams** [comment]
> `rustup toolchain install 1.54` might do the trick for you

ok great will try that thanks. :)

**Comment by dellams** [comment]
Installed it but still getting the same error, any more ideas? Thanks.

**Comment by Connoropolous** [comment]
Try doing:
`cargo install --path crates/holochain -Z resolver`

relates to what appears to be an unstable feature...
https://doc.rust-lang.org/cargo/reference/unstable.html?highlight=unstable#unstable-features

**Comment by dellams** [comment]
Ok thanks, I tried that but I got this error:

```
C:\Users\david\holochain>cargo install --path crates/holochain -Z resolver
error: the `-Z` flag is only accepted on the nightly channel of Cargo, but this is the `stable` channel
See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
```

**Comment by dellams** [comment]
Shall I try without the Z flag?

**Comment by dellams** [comment]
Without the Z flag I get a different error:

```
C:\Users\david\holochain>cargo install --path crates/holochain resolver
error: could not find `resolver` in C:\Users\david\holochain\crates\holochain with version `*`
```

Any ideas thanks? :)

**Comment by Connoropolous** [comment]
> Installed it but still getting the same error, any more ideas? Thanks.

After you installed it, did you switch to it? I can't remember if it switches automatically. If it was already switched and didn't work, then try nightly rust.

rustup toolchain install nightly
rustup default nightly 


Then run the original build or install commands

**Comment by dellams** [comment]
Ok, thanks, done that and run the `cargo install --path crates/holochain `command again and got a new error:

```
C:\Users\david\holochain>cargo install --path crates/holochain
  Installing holochain v0.0.102 (C:\Users\david\holochain\crates\holochain)
    Updating crates.io index
error: failed to compile `holochain v0.0.102 (C:\Users\david\holochain\crates\holochain)`, intermediate artifacts can be found at `C:\Users\david\holochain\target`

Caused by:
  failed to download `bit-set v0.2.0`

Caused by:
  unable to get packages from source

Caused by:
  failed to parse manifest at `C:\Users\david\.cargo\registry\src\github.com-1ecc6299db9ec823\bit-set-0.2.0\Cargo.toml`

Caused by:
  dependency (bit-vec) specified without providing a local path, Git repository, or version to use.

C:\Users\david\holochain>
```

Do I need to manually download bit-set v0.2.0 then?

Thanks. :)

**Comment by dellams** [comment]
Or issue with the Cargo.toml file?

**Comment by Connoropolous** [comment]
I'm not sure about it...

However, since version 0.0.102 has been published... could you try `cargo install holochain`?

**Comment by dellams** [comment]
Sure, when was it published? 🙂

**Comment by Connoropolous** [comment]
11 days ago https://crates.io/crates/holochain/0.0.102

If that won't work then we can log an issue separate to this one with the team and get it bumped for review cause that would seem to affect many people potentially

**Comment by dellams** [comment]
I ran `cargo install holochain` and got the same error:

```
C:\Users\david\holochain>cargo install holochain
    Updating crates.io index
  Downloaded holochain v0.0.102
  Downloaded 1 crate (279.9 KB) in 2.09s
  Installing holochain v0.0.102
error: failed to compile `holochain v0.0.102`, intermediate artifacts can be found at `C:\Users\david\AppData\Local\Temp\cargo-install7cpdpQ`

Caused by:
  failed to download `bit-set v0.2.0`

Caused by:
  unable to get packages from source

Caused by:
  failed to parse manifest at `C:\Users\david\.cargo\registry\src\github.com-1ecc6299db9ec823\bit-set-0.2.0\Cargo.toml`

Caused by:
  dependency (bit-vec) specified without providing a local path, Git repository, or version to use.
```

Any other ideas? :) Thanks

**Comment by Connoropolous** [comment]
Let me see if I hit that, installing for macos

**Comment by dellams** [comment]
This is now redundant since latest support is now in this issue: https://github.com/holochain/holochain/issues/2577
